src/system.rs

```diff
@@ -31,3 +127,88 @@ mod unit_tests {
         assert_eq!(uptime_internal("12489513.08 22906637.29\n"), 12489513);
     }
 }
+
+/// Returns information about virtual memory usage
+///
+/// `/proc/meminfo` contains the virtual memory statistics
+pub fn virtual_memory() -> Option<VirtualMemory> {
+    let data = read_file(Path::new("/proc/meminfo")).unwrap();
+    let mem_info = make_map(&data);
+
+    let mut virtual_memory: VirtualMemory = Default::default();
+
+    virtual_memory.total = get_deref!(mem_info, "MemTotal:");
+    virtual_memory.free = get_deref!(mem_info, "MemFree:");
+    virtual_memory.buffers = get_deref!(mem_info, "Buffers:");
+    virtual_memory.cached = get_deref!(mem_info, "Cached:");
+    virtual_memory.active = get_deref!(mem_info, "Active:");
+    virtual_memory.inactive = get_deref!(mem_info, "Inactive:");
+
+    // MemAvailable was introduced in kernel 3.14. The original psutil computes it if it's not
+    // found, but since 3.14 has already reached EOL, let's assume that it's there.
+    virtual_memory.available = get_deref!(mem_info, "MemAvailable:");
+
+    // Shmem was introduced in 2.6.19
+    virtual_memory.shared = get_deref!(mem_info, "Shmem:");
+
+    virtual_memory.used = virtual_memory.total - virtual_memory.free - virtual_memory.cached -
+        virtual_memory.buffers;
+    virtual_memory.percent = (virtual_memory.used as f32 / virtual_memory.total as f32) * 100.0;
+
+    Some(virtual_memory)
+}
+
+/// Returns information about swap memory usage
+///
+/// `/proc/meminfo` and `/proc/vmstat` contains the information
+pub fn swap_memory() -> Option<SwapMemory> {
+    let data = read_file(Path::new("/proc/meminfo")).unwrap();
+    let swap_info = make_map(&data);
+
+    let vmstat = read_file(Path::new("/proc/vmstat")).unwrap();
+    let vmstat_info = make_map(&vmstat);
+
+    let mut swap_memory: SwapMemory = Default::default();
+
+    swap_memory.total = get_deref!(swap_info, "SwapTotal:");
+    swap_memory.free = get_deref!(swap_info, "SwapFree:");
+    swap_memory.sin = get_deref!(vmstat_info, "pswpin");
+    swap_memory.sout = get_deref!(vmstat_info, "pswpout");
+    swap_memory.used = swap_memory.total - swap_memory.free;
+    swap_memory.percent = (swap_memory.used as f32 / swap_memory.total as f32) * 100.0;
+
+    Some(swap_memory)
+}
+
+fn get_multiplier(fields: &mut Vec<&str>) -> Option<u64> {
+    if let Some(ext) = fields.pop() {
+        let multiplier = match ext {
+            "kB" => Some(1024),
+            _ => None,
+        };
+        fields.push(ext);
+
+        multiplier
+    } else {
+        None
+    }
+}
+
+fn make_map(data: &str) -> HashMap<&str, u64> {
+    let lines: Vec<&str> = data.lines().collect();
+    let mut map = HashMap::new();
+
+    for line in lines {
+        let mut fields: Vec<&str> = line.split_whitespace().collect();
+        let key = fields[0];
+        let mut value = fields[1].parse::<u64>().unwrap();
```

Comment:
I'm slightly nervous of unwrap() here - if a new Linux version added a non-integer field to /proc/meminfo, this would make programs start failing without many diagnostics.

Not sure what's better:

- fail the whole call with an error if we see an unparseable field
- parse the parseable fields, and silently ignore unparseable ones
- parse the parseable fields, and add a dependency on https://docs.rs/log/0.4.1/log/ so we can log the unparseable ones

Reply:
I think making it an all or nothing function or logging the ones that can't be parsed are good options. Between those two though, I'm leaning towards the first.

If the kernel does add some non-parseable fields that end up being needed, I figure it would be best to fail early and fail hard.

Reply:
Yes, that seems reasonable - probably better to take the strictest approach (and then it's OK to relax it if need be) than to take a more relaxed approach and brak back-compatibility later.